### PR TITLE
more changes

### PR DESCRIPTION
**Pull request by chsievers**
Some proposed changes to the first four chapters and the accompanying code.

Please also have a look at my 'remarks' branch where I marked lines that I felt might be improved without having concrete suggestions. I plan to continue committing remarks to that branch, so you might want to track it.

### CODE REVIEW

File: 002_parsers.md [hunk 1]
```diff
@@ -70,7 +71,7 @@ it over the result of second parse function. Since the parser operation yields a
 list of tuples, composing a second parser function simply maps itself over the
 resulting list and concat's the resulting nested list of lists into a single
 flat list in the usual list monad fashion. The unit operation injects a single
-pure value into the parse stream.
+pure value as result without reading from the parse stream.
```
Comment (sdiehl) [inline 002_parsers.md]: Can't quite parse this. Should it be?

>The unit operation injects a single as the result, without reading from the parse stream.

File: chapter3/calc/Eval.hs [hunk 2]
```diff
@@ -34,6 +34,6 @@ nf :: Expr -> Expr
 nf x = fromMaybe x (nf <$> eval' x)
 
 eval :: Expr -> Maybe Expr
-eval t = case isVal (nf t) of
-  True  -> Just (nf t)
-  False -> Nothing -- term is "stuck"
+eval t = case nf t of
+  nft | isVal nft -> Just nft
+      | otherwise -> Nothing -- term is "stuck"
```
Comment (sdiehl) [inline chapter3/calc/Eval.hs]: I guess the most idiomatic Haskell thing here would be an if/then statement.

```haskell
eval t = 
  if isVal (nf t)
  then Just (nf t)
  else Nothing -- term is "stuck"
```

File: chapter3/parsec.hs [hunk 3]
```diff
@@ -54,9 +54,9 @@ failure = Parser (\cs -> [])
 
 option :: Parser a -> Parser a -> Parser a
 option  p q = Parser $ \s ->
-  case parse (mplus p q) s of
-    []     -> []
-    (x:xs) -> [x]
+  case parse p s of
```
Comment (sdiehl) [inline chapter3/parsec.hs]: What was wrong here?

**Comment by sdiehl** [comment]
Thanks again, I made a few minor nits, but I think I can merge this pretty quick so long as I understand what the changes to the ``option`` function involved. Was there a bug?

File: 002_parsers.md (hunk 1 above)
Comment (chsievers) [inline 002_parsers.md]: Yes, that's fine.
Original formulation sounded like we're manipulating the input.

File: chapter3/parsec.hs (hunk 3 above)
Comment (chsievers) [inline chapter3/parsec.hs]: ``mzero`` wasn't a neutral element for ``(<|>)`` - see message of commit 1170925.
It makes no difference as long as all our parsers return at most one parse.

**Comment by chsievers** [comment]
Don't know about ``option``, but ``(<|>)`` didn't seem to satisfy its specification.

File: chapter3/calc/Eval.hs (hunk 2 above)
Comment (chsievers) [inline chapter3/calc/Eval.hs]: That again recomputes the normal form.
I'd suggest:
```haskell
eval t =
  let nft = nf t in
    if isVal nft
    then nft
    else Nothing -- term is "stuck"
```
Before I tried to keep the style and line count.
(Forgot to click on "Comment" earlier...)

Comment (sdiehl) [inline chapter3/calc/Eval.hs]: Ahh, I see. The version you proposed the line numbers (which are kind of a pain to change) so we'll just go with the guarded version.

File: 002_parsers.md (hunk 1 above)
Comment (sdiehl) [inline 002_parsers.md]: Fixed in fdf6748.

**Pull request merged**